### DISCUSSION TITLE
http://lotabout.github.io/2015/write-a-C-interpreter-3/

### DISCUSSION THREAD

**Discussion by lotabout** [lotabout/lotabout.github.io#12] [Announcements]
# http://lotabout.github.io/2015/write-a-C-interpreter-3/
手把手教你构建 C 语言编译器（3）- 词法分析器
http://lotabout.github.io/2015/write-a-C-interpreter-3/

**Comment by lotabout** [comment]
由 shellRaining 发布于 2023-05-19T01:42:44Z：

本章节要注意在 main 函数 malloc 一个 symbol 表

**Comment by lotabout** [comment]
由 silentdxx 发布于 2024-04-24T02:16:09Z：

我来解释下对 **token_val = token_val \* 16 + (token \& 15) + (token \>= 'A' ? 9 : 0);** 16进制字符串转化数字的理解:

1. 传统的转换方式上，要分别判断当前字符是属于a\~f、A\~F或0\~9三种情况哪儿一种，可以把16进制字符串实现为下面的代码：  

`token_val = token_val * 16 + (token >= 'a' ? token - 'a' + 10 : token >= 'A' ? token - 'A' + 10 : token - '0');`  

更加直观的计算方式如下:  

` if (token >= '0' && token <= '9') {`  
` return token - '0';`  
` } else if (token >= 'a' && token <= 'f') {`  
` return 10 + token - 'a';`  
` } else if (token >= 'A' && token <= 'F') {`  
` return 10 + token - 'A';`  
` }`  

2. **token_val = token_val \* 16 + (token \& 0xF) + (token \>= 'A' ? 9 : 0);** 使用了特殊技巧来获取当前字符的数值，关键在于对**(token \& 15) + (token \>= 'A' ? 9 : 0)** 的理解，我们先仔细看一下ascii表:  
[https://uploads.disquscdn.c...](https://uploads.disquscdn.com/images/ebb080b29ad6c92b4abf599c2810c5334fc49ec281816ed15d34eeaf5bf76d33.png "https://uploads.disquscdn.com/images/ebb080b29ad6c92b4abf599c2810c5334fc49ec281816ed15d34eeaf5bf76d33.png")

1）可以看出token \& 15仅仅计算出当前字符的低4位，正好等于0\~9位  
2）当token \>= 'A'，此时为a\~f、A\~F这两种情况，我们看到a\~f的16进制值为41\~46，A\~F的16进制值为61\~66，也就是a\~f与A\~F的低4位值都等于1\~6，所以要加上9这个基数，也就是10\~15，这个想法简直太棒了！

**Comment by LeeeeeeM** [comment]
你这节写的跳跃性太大
